### ISSUE TITLE
System crash when using file browser to view docker volumes directory

### ISSUE THREAD

**Issue by holocronology** [ravenclaw900/DietPi-Dashboard#513]
#### Required Information
- DietPi version | 8.15.1 (beta) -- also experienced this on current release version
- Distro version | bullseye 0
- Kernel version | `Linux DietPi 5.15.84-v8+ #1613 SMP PREEMPT Thu Jan 5 12:03:08 GMT 2023 aarch64 GNU/Linux`
- SBC model | RPi 3 Model B+ (aarch64)
- Power supply used | 5V 5A
- SD card used | Samsung Evo Pro

#### Additional Information (if applicable)
- Software title | dietpi-dashboard, docker
- Was the software title installed freshly or updated/migrated? fresh
- Can this issue be replicated on a fresh installation of DietPi? yes

- Bug report ID | 0d4cdd01-2b81-40d6-884e-b157253dfbb8

#### Steps to reproduce
1. Access dietpi-dashboard
2. Select "file manager"
3. navigate to /mnt/dietpi_userdata/docker-data
4. click on "volumes" folder
5. system hangs, physical power off/reboot required

#### Expected behaviour

- It should open the "volumes" directory the same as other folders and not crash/hang system

#### Actual behaviour

- when you click on the "volumes' directory the system crashes/hangs/is extremely slow

#### Extra details

- when clicking on volumes directory it does not even open:
![image](https://user-images.githubusercontent.com/65265114/224486083-d3e2815e-258c-493a-b3d6-77df97fbd10f.png)

**Comment by MichaIng** [comment]
Docker volumes are `overlayfs` mounts, but the parent directory isn't anything special, AFAIK. Let me have a look.

**Comment by MichaIng** [comment]
It is a normal directory, but probably this block device file causes issues with the way the dashboard (respectively the Rust create/method used) obtains file info:
```console
root@VM-Bullseye:/mnt/dietpi_userdata/docker-data/volumes# ls -l
brw------- 1 root root 8, 1 Mar 11 14:34 backingFsBlockDev
```

**Comment by MichaIng** [comment]
@holocronology 
You cannot even access via SSH when it hangs there? Can you open an SSH terminal and run this command:
```sh
journalctl -f
```
then replicate the issue. The command keeps printing system logs, including kernel errors and such, so probably that way we are able to catch some useful info.

**Comment by holocronology** [comment]
> @holocronology You cannot even access via SSH when it hangs there? Can you open an SSH terminal and run this command:
> 
> ```shell
> journalctl -f
> ```
> 
> then replicate the issue. The command keeps printing system logs, including kernel errors and such, so probably that way we are able to catch some useful info.

No, I cannot SSH after this condition is triggered. In fact, I just accidentally clicked on the `var` directory in the file browser and replicated the same issue. 

The only way out that I found was to remove power.

**Comment by holocronology** [comment]
> @holocronology You cannot even access via SSH when it hangs there? Can you open an SSH terminal and run this command:
> 
> ```shell
> journalctl -f
> ```
> 
> then replicate the issue. The command keeps printing system logs, including kernel errors and such, so probably that way we are able to catch some useful info.

I let it sit and it eventually spat out this: https://pastebin.com/0ZK22SzK

**Comment by MichaIng** [comment]
The OOM (Out-Of-Memory) killer became active and killed also essential system services.

@ravenclaw900 
I remember there was such an issue already with block device files, as the dashboard/file browser seems to read the whole device into memory or so? Probably fixed in `main`, requiring a stable release?

@holocronology 
Can you try to reinstall and select nightly build?
```sh
dietpi-software reinstall 200
```

**Comment by ravenclaw900** [comment]
All the dashboard is doing to get directories is calling `tokio::fs::read_dir`, which itself just runs `std::fs::read_dir` on a thread pool. The only way I can think of it running out of memory is an insane number of files in one flat directory, or some filesystem shenanigans that cause an infinite loop. Let me add some more debug messages to the nightly to see if we can figure out what's going on.

**Comment by ravenclaw900** [comment]
@holocronology
Try the nightly build with debug logging enabled in the config file

**Comment by holocronology** [comment]
> @holocronology Try the nightly build with debug logging enabled in the config file

Can do! Please remind me where to find the config file.

So far, seems to be doing the same.

**Comment by ravenclaw900** [comment]
It should be located at `/opt/dietpi-dashboard/config.toml`.

**Comment by holocronology** [comment]
> /opt/dietpi-dashboard/config.toml

Thanks. Now where is the log? :-)

**Comment by MichaIng** [comment]
```sh
journalctl -u dietpi-dashboard
```

**Comment by holocronology** [comment]
@ravenclaw900 

Here: https://pastebin.com/yaWLvNqR

**Comment by holocronology** [comment]
Also, this: https://pastebin.com/zq3J0tSV

![image](https://user-images.githubusercontent.com/65265114/224553637-e9c9b029-dceb-4a1f-a223-ddfe75cb0911.png)

**Comment by ravenclaw900** [comment]
Okay, so the dashboard is trying to read something that's not a directory, file, or symlink (thus causing my debug message code to crash, whoops). What's the contents of the `/mnt/dietpi_userdata/docker-data` directory?
```
ls -l /mnt/dietpi-userdata/docker-data
```

**Comment by holocronology** [comment]
huh...

`ls: cannot access '/mnt/dietpi-userdata/docker-data': No such file or directory`

**Comment by holocronology** [comment]
This:
`
root@DietPi:~# ls -l /mnt/dietpi_userdata/docker-data
total 44
drwx--x--x 4 root root 4096 Mar 11 07:37 buildkit
drwx--x--- 2 root root 4096 Mar 11 07:37 containers
-rw------- 1 root root   36 Mar 11 07:37 engine-id
drwx------ 3 root root 4096 Mar 11 07:37 image
drwxr-x--- 3 root root 4096 Mar 11 07:37 network
drwx--x--- 3 root root 4096 Mar 12 11:47 overlay2
drwx------ 4 root root 4096 Mar 11 07:37 plugins
drwx------ 2 root root 4096 Mar 12 11:47 runtimes
drwx------ 2 root root 4096 Mar 11 07:37 swarm
drwx------ 2 root root 4096 Mar 12 11:47 tmp
drwx-----x 2 root root 4096 Mar 12 11:47 volumes
`

The the fail is on `/mnt/dietpi_userdata/docker-data/volumes`, which shows:
`root@DietPi:~# ls -l /mnt/dietpi_userdata/docker-data/volumes
total 24
brw------- 1 root root 179, 2 Mar 12 11:47 backingFsBlockDev
-rw------- 1 root root  32768 Mar 12 11:47 metadata.db
`

**Comment by ravenclaw900** [comment]
Okay, so I bet the problem is that, when trying to guess the file type, the dashboard ends up reading all of `backingFsBlockDev` into memory (pretty much what @MichaIng guessed earlier). Let me add a few checks to make sure it's reading a file before trying to guess the file type.

**Comment by holocronology** [comment]
> Okay, so I bet the problem is that, when trying to guess the file type, the dashboard ends up reading all of `backingFsBlockDev` into memory (pretty much what @MichaIng guessed earlier). Let me add a few checks to make sure it's reading a file before trying to guess the file type.

Seems to have the same results on the `/var` directory also:
root@DietPi:~# ls -l /var
total 1115176
drwxr-xr-x  2 root root        4096 Mar 12 01:25 backups
drwxr-xr-x  7 root root        4096 Mar 11 08:32 cache
drwxr-xr-x 15 root root        4096 Mar 11 18:22 lib
drwxrwsr-x  2 root staff       4096 Dec  9 14:15 local
lrwxrwxrwx  1 root root           9 Feb 13 15:21 lock -> /run/lock
drwxr-xr-x  6 root root         240 Mar 11 08:36 log
drwxrwsr-x  2 root mail        4096 Feb 13 15:34 mail
drwxr-xr-x  2 root root        4096 Feb 13 15:21 opt
lrwxrwxrwx  1 root root           4 Feb 13 15:21 run -> /run
drwxr-xr-x  3 root root        4096 Feb 13 15:23 spool
-rw-------  1 root root  1141899264 Feb 13 15:37 swap
drwxrwxrwt  3 root root        4096 Mar 12 12:09 tmp
drwxr-xr-x  3 root root        4096 Mar 11 08:37 www

**Comment by ravenclaw900** [comment]
Please test the latest nightly again, it should resolve the problem with the docker volumes directory. The issue with `/var` might be a different problem, since the only special files there are symlinks, and they only link to directories. Regardless, see if the nightly fixes that too.

**Comment by holocronology** [comment]
That works. I am able to get to mnt/.../docker-data/volumes.

going to /var did not work, same old results.

**Comment by ravenclaw900** [comment]
Can you print the debug log showing what happens if you try to navigate to `/var`?

**Comment by holocronology** [comment]
yup: https://pastebin.com/eWRt3PV6

**Comment by ravenclaw900** [comment]
From the logs, it looks like everything is fine, but I'm going to guess that it's timing out trying to read `/var/swap`. Since this actually is a normal file, the previous check doesn't apply here. Should be fixed with #516.
